src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp

```diff
@@ -177,137 +84,37 @@ static void havoc_if_valid(
     skip_program.add(goto_programt::make_skip(source_location_no_checks));
 
   dest.add(goto_programt::make_goto(
-    skip_target, not_exprt{target_valid_var}, source_location_no_checks));
+    skip_target, not_exprt{car.valid_var()}, source_location_no_checks));
 
-  if(target.id() == ID_pointer_object)
+  if(car.target().id() == ID_pointer_object)
   {
     // OTHER __CPROVER_havoc_object(target_snapshot_var)
-    codet code(ID_havoc_object, {target_snapshot_var});
+    codet code(ID_havoc_object, {car.lower_bound_var()});
     const auto &inst =
       dest.add(goto_programt::make_other(code, source_location));
     inst->source_location_nonconst().set_comment(tracking_comment);
   }
   else
   {
-    // ASSIGN *target_snapshot_var = nondet()
-    side_effect_expr_nondett nondet(target.type(), source_location);
+    // ASSIGN *(target.type() *)__car_lb = nondet(car.target().type())
+    const auto &target_type = car.target().type();
+    side_effect_expr_nondett nondet(target_type, source_location);
     const auto &inst = dest.add(goto_programt::make_assignment(
-      dereference_exprt{target_snapshot_var}, nondet, source_location));
+      dereference_exprt{typecast_exprt::conditional_cast(
+        car.lower_bound_var(), pointer_type(target_type))},
+      nondet,
+      source_location));
```

Comment:
I think we might be better off using `nondet_initializer` here for that would produce nondets for individual fields (based on the type). `goto-symex` could then take each of these in its field-sensitive encoding.

Reply:
The latter will use `side_effect_expr_nondett` by either returning exactly such (for all basic types), or by creating a compound object that in turn contains `side_effect_expr_nondett` (for example a struct expression with a nondet for each member).